Add DAM image block

johnnyomair opened the pull request:
We usually use a `OneOfBlock` to create an image block that supports both pixel and SVG images. This lead to UX problems when using the image block, for instance SVGs could only be selected when the SVG image block was active. This change introduces a special DAM image block to resolve the UX problems. The block uses a `OneOfBlock` under the hood but provides a custom Admin component which allows users to choose any image when empty, and renders the respective child block's Admin component based on the selected file.

**Before**

https://user-images.githubusercontent.com/48853629/178985551-46baa71c-378e-437d-b7a4-b5b4da02a872.mov

**After**

https://user-images.githubusercontent.com/48853629/178987123-d9dd27dd-cd0d-4bab-bc7c-3d1f3119dc52.mov

nsams commented:
Does crop settings (or other settings) of the PixelImage block still work, or don't we need those?

johnnyomair commented:
> Does crop settings (or other settings) of the PixelImage block still work, or don't we need those?

Yes, cropping still works. However, cropping is lost when switching to the SVG image block, as the user can't reselect the already selected pixel image block, but choose a new pixel image instead.

thomasdax98 commented:
I noticed 2 things:

1) In the Comet Demo project, there are now two blocks called "Image" in the select menu. 

<img width="337" alt="Bildschirmfoto 2022-07-18 um 09 47 47" src="https://user-images.githubusercontent.com/13380047/179467855-e483bbed-6b49-4dd1-b1f3-54fd8059d288.png">

Will the new image block replace the old one? Is their data compatible?
Or will they exist in parallel?

2) The name of the block changes depending on which type of image is selected.

https://user-images.githubusercontent.com/13380047/179468160-74027b26-1134-4c7a-99d0-6d3d0139795d.mov

Is this the desired behavior?
I guess it could be confusing for the user because once a SVG is selected, the block is called SVG.

nsams commented:
> Yes, cropping still works. However, cropping is lost when switching to the SVG image block, as the user can't reselect the already selected pixel image block, but choose a new pixel image instead.

I don't understand this code by reading it :)

Are other settings (in image block) are also supported? Alt-Text or other SEO stuff for example. Or a (web specific) caption - or should this be at the OneOf level - to have it for both image types?

johnnyomair commented:
> I don't understand this code by reading it :)

Do you mean the _sentence_, or the code of the Admin component? 😁 

> Are other settings (in image block) are also supported? Alt-Text or other SEO stuff for example. Or a (web specific) caption - or should this be at the OneOf level - to have it for both image types?

We've implemented features such as captions, aspect ratios etc. by creating a "ConfigurableImageBlock" in the past. The PixelImageBlock and SvgImageBlock don't support such features at the moment.

johnnyomair commented:
What do you think about the approach (using a OneOfBlock) in general? Would it be better to just implement it as a completely separate block?

nsams commented:
> Do you mean the sentence, or the code of the Admin component? grin

the code of the admin component....

> 
We've implemented features such as captions, aspect ratios etc. by creating a "ConfigurableImageBlock" in the past. The PixelImageBlock and SvgImageBlock don't support such features at the moment.

:+1:

> What do you think about the approach (using a OneOfBlock) in general? Would it be better to just implement it as a completely separate block?

:+1: good idea!

johnnyomair commented:
I've created a completely separate DAM image block: b189af18566a8069f53570a1e824a19b5f85189b

nsams commented:
> I've created a completely separate DAM image block

why did you change that now?

johnnyomair commented:
> why did you change that now?

It thought you meant that creating a completely separate block is a good idea:

> >What do you think about the approach (using a OneOfBlock) in general? Would it be better to just implement it as a completely separate block?
>
> 👍 good idea!

Did you mean that using an OneOfBlock is a good idea?

nsams commented:
yes - that was a misunderstanding - sorry!

Now that you implemented both - what do you prefer?

johnnyomair commented:
> Now that you implemented both - what do you prefer?

I prefer the approach of having a completely separate block. Sure, some parts are duplicated, but it's less magic. Kind of reminds me of our discussion whether we should implement the ColumnsBlock using a ListBlock or as a standalone block.

The only thing I liked more about the OneOfBlock approach was the possibility to use the new block as a drop-in replacement. However, I've not actually tried that, and I believe there would be issues with different block versions, as the image block has been migrated once before (in Demo and in some of our projects).

nsams commented on packages/api/cms-api/src/dam/images/images.service.ts:
```diff
@@ -22,8 +22,13 @@ export class ImagesService {
+        const usedCropArea = cropArea ?? file.image?.cropArea;
+
+        if (usedCropArea === undefined) {
+            return undefined;
```
when is usedCropArea undefined?

nsams commented:
> I prefer the approach of having a completely separate block.

your decision :+1: 

What could be confusing is that in site `DamImageBlock` uses `PixelImageBlock` and `SvgImageBlock` although those are not existing anymore as block. If you move that component into the library you would hide that....

johnnyomair commented:
@nsams I've reverted to the `OneOfBlock` approach once again 😅 It eliminates any unnecessary duplication and actually isn't magic at all.

The completely separate DAM image block introduced much complexity in unwanted places, for instance the `urlTemplate` had to be `undefined` for SVG images and a `string` for pixel images.

I've reworked the custom Admin component to make it more readable (in my opinion).
